### DISCUSSION TITLE
Upgrading to 1.0.0. Here's some tips.

### DISCUSSION THREAD

**Discussion by KingWebsites** [opensource9ja/danfojs#348] [Show and tell]
Hi, I've just upgraded to 1.0.0 and I thought I'd list a couple of the changes I did to get my project working again. Hopefully it will save some of you some time. Please feel free to list more below.

Please read the [official migration guide here](https://danfo.jsdata.org/examples/migrating-to-the-stable-version-of-danfo.js), which has lots of advice.

1. camelCase. These were a few of the things I had to change:

Concatenating: "df_list" to "dfList"
```
const newDF = dfd.concat({ df_list: [dataFrameA.column("COL A"), dataFrameB.column("COL B")], axis: 1}) // old
const newDF = dfd.concat({ dfList: [dataFrameA.column("COL A"), dataFrameB.column("COL B")], axis: 1}) // new
```
*note. I'm not sure why my "axis: 1" worked before.
**note. Concat now removes index names so you have to store and add them back in with setIndex.

Counts of nUnique items: "nunique" to "nUnique" & "value_counts()" to "valueCounts()"
```
const count = dataFrame.nunique(0).value_counts().values[0]; // old
const count = dataFrame.nUnique(1).valueCounts().values[0]; // new
```
*note. axis value also changed from 0 to 1.

setIndex
```
dataFrame.set_index({
dataFrame.setIndex({
```

2. loc column and row arguments don't accept basic variables like before. They need to be wrapped as a "template literal".

```
const date = "2021-12-14"
const rowDataFrame = dataFrame.loc({ rows: [date] }); // old
const rowDataFrame = dataFrame.loc({ rows: [`"${date}"`] }); // new
```

```
const date = "2021-12-14"
const cellValue = dataFrame.loc({ rows: [date], columns: ["Col B"]} ).values.flat()[0]; // old
const cellValue = dataFrame.loc({ rows: [`"${date}"`], columns: ["Col B"]} ).values.flat()[0]; // new
```
*note. Should I be using a different method just to get the single value of cell X/Y?

**Comment by steveoni** [comment]
@KingWebsites thanks for the notice. what do you mean by this `*note. I'm not sure why my "axis: 1" worked before.` ?

and for the index name, it will be added back

**Reply by KingWebsites** [reply]
Hi.

Re point 1, I was comparing my new commit and I noticed that I hadn't changed the axis value yet it still worked. Strange.

As for the index being removed when you concat maybe this is a bug? The row names get lost when I concat but they never used to. Here's some some test code which shows the problem:

```
let data_a = {
	"Name": ["Apples", "Mango", "Banana", "Pear"],
	"Count": [21, 5, 30, 10],
	"Price": [200, 300, 40, 250]
}
let df_a = new dfd.DataFrame(data_a, { index: ["Row A", "Row B", "Row C", "Row D"] })
df_a.print();

let data_b = {
	"Colour": ["Green", "Mango", "Yellow", "Green"],
	"Taste": [8, 9, 7, 6]
}
let df_b = new dfd.DataFrame(data_b, { index: ["Row A", "Row B", "Row C", "Row D"] })
df_b.print();

let df_joined = dfd.concat({ dfList: [df_a, df_b], axis: 1})
df_joined.print();
```
```

╔════════════╤═══════════════════╤═══════════════════╤═══════════════════╗
║            │ Name              │ Count             │ Price             ║
╟────────────┼───────────────────┼───────────────────┼───────────────────╢
║ Row A      │ Apples            │ 21                │ 200               ║
╟────────────┼───────────────────┼───────────────────┼───────────────────╢
║ Row B      │ Mango             │ 5                 │ 300               ║
╟────────────┼───────────────────┼───────────────────┼───────────────────╢
║ Row C      │ Banana            │ 30                │ 40                ║
╟────────────┼───────────────────┼───────────────────┼───────────────────╢
║ Row D      │ Pear              │ 10                │ 250               ║
╚════════════╧═══════════════════╧═══════════════════╧═══════════════════╝
╔════════════╤═══════════════════╤═══════════════════╗
║            │ Colour            │ Taste             ║
╟────────────┼───────────────────┼───────────────────╢
║ Row A      │ Green             │ 8                 ║
╟────────────┼───────────────────┼───────────────────╢
║ Row B      │ Mango             │ 9                 ║
╟────────────┼───────────────────┼───────────────────╢
║ Row C      │ Yellow            │ 7                 ║
╟────────────┼───────────────────┼───────────────────╢
║ Row D      │ Green             │ 6                 ║
╚════════════╧═══════════════════╧═══════════════════╝
╔════════════╤═══════════════════╤═══════════════════╤═══════════════════╤═══════════════════╤═══════════════════╗
║            │ Name              │ Count             │ Price             │ Colour            │ Taste             ║
╟────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────╢
║ 0          │ Apples            │ 21                │ 200               │ Green             │ 8                 ║
╟────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────╢
║ 1          │ Mango             │ 5                 │ 300               │ Mango             │ 9                 ║
╟────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────╢
║ 2          │ Banana            │ 30                │ 40                │ Yellow            │ 7                 ║
╟────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────╢
║ 3          │ Pear              │ 10                │ 250               │ Green             │ 6                 ║
╚════════════╧═══════════════════╧═══════════════════╧═══════════════════╧═══════════════════╧═══════════════════╝
```

**Reply by risenW** [reply]
> Hi.
> 
> Re point 1, I was comparing my new commit and I noticed that I hadn't changed the axis value yet it still worked. Strange.
> 
> As for the index being removed when you concat maybe this is a bug? The row names get lost when I concat but they never used to. Here's some some test code which shows the problem:
> 
> ```
> let data_a = {
> 	"Name": ["Apples", "Mango", "Banana", "Pear"],
> 	"Count": [21, 5, 30, 10],
> 	"Price": [200, 300, 40, 250]
> }
> let df_a = new dfd.DataFrame(data_a, { index: ["Row A", "Row B", "Row C", "Row D"] })
> df_a.print();
> 
> let data_b = {
> 	"Colour": ["Green", "Mango", "Yellow", "Green"],
> 	"Taste": [8, 9, 7, 6]
> }
> let df_b = new dfd.DataFrame(data_b, { index: ["Row A", "Row B", "Row C", "Row D"] })
> df_b.print();
> 
> let df_joined = dfd.concat({ dfList: [df_a, df_b], axis: 1})
> df_joined.print();
> ```
> 
> ```
> 
> ╔════════════╤═══════════════════╤═══════════════════╤═══════════════════╗
> ║            │ Name              │ Count             │ Price             ║
> ╟────────────┼───────────────────┼───────────────────┼───────────────────╢
> ║ Row A      │ Apples            │ 21                │ 200               ║
> ╟────────────┼───────────────────┼───────────────────┼───────────────────╢
> ║ Row B      │ Mango             │ 5                 │ 300               ║
> ╟────────────┼───────────────────┼───────────────────┼───────────────────╢
> ║ Row C      │ Banana            │ 30                │ 40                ║
> ╟────────────┼───────────────────┼───────────────────┼───────────────────╢
> ║ Row D      │ Pear              │ 10                │ 250               ║
> ╚════════════╧═══════════════════╧═══════════════════╧═══════════════════╝
> ╔════════════╤═══════════════════╤═══════════════════╗
> ║            │ Colour            │ Taste             ║
> ╟────────────┼───────────────────┼───────────────────╢
> ║ Row A      │ Green             │ 8                 ║
> ╟────────────┼───────────────────┼───────────────────╢
> ║ Row B      │ Mango             │ 9                 ║
> ╟────────────┼───────────────────┼───────────────────╢
> ║ Row C      │ Yellow            │ 7                 ║
> ╟────────────┼───────────────────┼───────────────────╢
> ║ Row D      │ Green             │ 6                 ║
> ╚════════════╧═══════════════════╧═══════════════════╝
> ╔════════════╤═══════════════════╤═══════════════════╤═══════════════════╤═══════════════════╤═══════════════════╗
> ║            │ Name              │ Count             │ Price             │ Colour            │ Taste             ║
> ╟────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────╢
> ║ 0          │ Apples            │ 21                │ 200               │ Green             │ 8                 ║
> ╟────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────╢
> ║ 1          │ Mango             │ 5                 │ 300               │ Mango             │ 9                 ║
> ╟────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────╢
> ║ 2          │ Banana            │ 30                │ 40                │ Yellow            │ 7                 ║
> ╟────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────┼───────────────────╢
> ║ 3          │ Pear              │ 10                │ 250               │ Green             │ 6                 ║
> ╚════════════╧═══════════════════╧═══════════════════╧═══════════════════╧═══════════════════╧═══════════════════╝
> ```

Yeah, that works because some methods (concat) already used the correct axis. 

Overall, thanks for sharing this!

**Reply by KingWebsites** [reply]
FYI: my concat code example above was to show how the row names (indexes) disappear when you join two dataframes together.

**Reply by risenW** [reply]
Oh I see, it's a bug. @steveoni is working on it 

Issue created here: https://github.com/javascriptdata/danfojs/issues/349. 

@KingWebsites Are you interested in adding this new feature?

**Comment by risenW** [comment]
Regarding your point on: 
> Should I be using a different method just to get the single value of cell X/Y?

I think we need to add a new method for that. Something like [iat](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iat.html) in pandas. I'll create an issue for it.
